src/Payments.sol

```diff
@@ -0,0 +1,612 @@
+     * @return settledUntil The final epoch that effectively got settled.
+     * @return note An optional note returned by the arbitrator.
+     */
+    function _settleSegment(
+        uint256 railId,
+        Rail storage rail,
+        uint256 epochStart,
+        uint256 epochEnd,
+        uint256 rate
+    ) 
+        internal 
+        returns (
+            uint256 settledAmount, 
+            uint256 settledUntil, 
+            string memory note
+        ) 
+    {
+        // Calculate the intended settlement amount for the time span.
+        uint256 duration = epochEnd - epochStart;
+        settledAmount = rate * duration;
+        settledUntil = epochEnd;
+
+        // If the rail has an assigned arbiter, call it for arbitration on this segment.
+        if (rail.arbiter != address(0)) {
+            IArbiter arbiter = IArbiter(rail.arbiter);
+            IArbiter.ArbitrationResult memory arbResult = arbiter.arbitratePayment(
+                railId,
+                settledAmount,
+                epochStart,
+                epochEnd
+            );
+
+            require(arbResult.approved, "failed to settle: arbiter refused payment");
+            require(
+                arbResult.settleUpto <= epochEnd,
+                "failed to settle: arbiter settled beyond segment end"
+            );
+            require(
+                arbResult.modifiedAmount <= rate * (arbResult.settleUpto - epochStart),
+                "failed to settle: arbiter modified amount exceeds maximum for settled duration"
+            );
+
+            // Adjust the amount and the final epoch according to the arbiter's decision.
+            settledUntil = arbResult.settleUpto;
+            settledAmount = arbResult.modifiedAmount;
+            note = arbResult.note;
+        }
+
+        // Revert if no progress was made (i.e., the final epoch did not advance).
+        require(settledUntil > epochStart, "failed to settle: no progress made in settlement");
+
+        return (settledAmount, settledUntil, note);
+    }
+
+    function _settleWithRateChanges(
+        uint256 railId,
+        Rail storage rail,
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 initialRate,
+        uint256 startEpoch,
+        uint256 targetEpoch
+    ) internal returns (uint256 totalSettled, uint256 finalEpoch, string memory note) {
+        totalSettled = 0;
+        uint256 currentEpoch = startEpoch;
+        uint256 activeRate = initialRate;
+        note = "";
+
+        while (currentEpoch <= targetEpoch) {
+            uint256 nextBoundary = targetEpoch;
+
+            // If there's an upcoming rate change, check if it applies within our current range.
+            if (!rateQueue.isEmpty()) {
+                RateChangeQueue.RateChange memory upcomingChange = rateQueue.peek();
+                bool isWithinRange = (
+                    upcomingChange.untilEpoch >= currentEpoch &&
+                    upcomingChange.untilEpoch <= targetEpoch
```

Comment:
This second condition won't be true if I'm trying to settle up to some historical epoch in the past.

Reply:
We need to remove this assert.

Reply:
Done.